### PR TITLE
[Offload] Add offload performance metrics for tiered storage.

### PR DESCRIPTION
**Pull request by frankxieke**
### Motivation
Currently, there is no offload metrics for tiered storage, so it is very hard for us to debug the performance issues. For example , we can not find why offload is slow or why read offload is slow.  For above reasons.  we need to add some  offload metrics for monitoring.  

### Modifications
Add metrics during offload procedure and read offload data procedure. Including offloadTime, offloadError, offloadRate, readLedgerLatency, writeStoreLatency, writeStoreError, readOffloadIndexLatency, readOffloadDataLatency, readOffloadRate, readOffloadError. 

### Documentation

need  documentation.

### CODE REVIEW

**Comment by Anonymitaet** [comment]
@frankxieke should we add metrics explanations here https://pulsar.apache.org/docs/en/next/reference-metrics/ ?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/LedgerOffloaderMXBean.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    /**
+     * The ledger offloader name.
+     *
+     * @return ledger offloader name.
+     */
+    String getName();
+
+
+    /**
+     * refresh offloader stats.
```
Comment (Anonymitaet) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/LedgerOffloaderMXBean.java:46]: Please capitalize the first word

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/LedgerOffloaderMXBean.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    /**
+     * Record the offload rate to storage.
+     *
+     * @return
+     */
+    Map<String, Rate> getOffloadRates();
+
+
+    /**
+     * Record the read ledger latency
```
Comment (Anonymitaet) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/LedgerOffloaderMXBean.java:77]: Pls add a period at the end of the sentence.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/LedgerOffloaderMXBean.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    /**
+     * Record the write to storage error count.
+     *
+     * @return
+     */
+    Map<String, Rate> getWriteToStorageErrors();
+
+    /**
+     * Record read offload index latency
```
Comment (Anonymitaet) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/LedgerOffloaderMXBean.java:98]: same comment as previous. pls check overall, thanks

**Review by Anonymitaet — CHANGES_REQUESTED** [review]

**Comment by Anonymitaet** [comment]
Confirmed w/ @frankxieke, he will add docs and ping me review after this PR is finalized.

**Review by michaeljmarshall — CHANGES_REQUESTED** [review]
@frankxieke - this will be a good addition to our metrics. Before we merge this PR, I would like to see these metrics behind a feature flag that defaults to `disabled` at the topic level. Perhaps we can use the config `exposeManagedLedgerMetricsInPrometheus`, which already exists, as the feature flag?

**Comment by michaeljmarshall** [comment]
@lhotari - I know you have contributed optimizations to the way Pulsar generates Prometheus metrics in the past. Would you mind taking a look at this PR?

**Review by codelipenghui — CHANGES_REQUESTED** [review]
Pulsar can enable/disable the namespace/topic level metrics. so we should aggregate for different levels, if namespace/topic level disabled we only expose the broker level metrics, if namespace enabled but topic level disabled, we only expose the namespace level aggregate value.

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlockAwareSegmentInputStreamImpl.java [hunk 4]
```diff
@@ -113,9 +123,12 @@ private int readEntries() throws IOException {
 
     private List<ByteBuf> readNextEntriesFromLedger(long start, long maxNumberEntries) throws IOException {
         long end = Math.min(start + maxNumberEntries - 1, ledger.getLastAddConfirmed());
+        long startTime = System.nanoTime();
         try (LedgerEntries ledgerEntriesOnce = ledger.readAsync(start, end).get()) {
-            log.debug("read ledger entries. start: {}, end: {}", start, end);
-
+            log.debug("read ledger entries. start: {}, end: {} cost {}", start, end, TimeUnit.NANOSECONDS.toMicros(System.nanoTime() - startTime));
```
[the remark this answers is not in the document]
Comment (frankxieke) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlockAwareSegmentInputStreamImpl.java:128]: ok

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/LedgerOffloaderMXBeanImpl.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    public static final long[] READ_ENTRY_LATENCY_BUCKETS_USEC = {500, 1_000, 5_000, 10_000, 20_000, 50_000, 100_000,
+            200_000, 1000_000};
+
+    private final String name;
+
+    private final ScheduledExecutorService scheduler;
+    private long refreshIntervalSeconds;
+
+    // offloadTimeMap record the time cost by one round offload
+    private final ConcurrentHashMap<String, Rate> offloadTimeMap = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (frankxieke) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/LedgerOffloaderMXBeanImpl.java:44]: There is only one global offloader.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsGenerator.java [hunk 6]
```diff
@@ -125,6 +135,24 @@ public static void generate(PulsarService pulsar, boolean includeTopicMetrics, b
...
+                if (ledgerOffloader != NullLedgerOffloader.INSTANCE && ledgerOffloader.getStats() != null) {
+                    String clusterName = pulsar.getConfiguration().getClusterName();
+                    List<Metrics> metrics = new LedgerOffloaderMetrics(pulsar, ledgerOffloader).generate();
+                    parseMetricsToPrometheusMetrics(metrics, clusterName, Collector.Type.GAUGE, stream);
+                }
+            } catch (Exception ex) {
+                log.error("generate ledger offloader metrics error", ex);
+            }
+        });
+    }
```
[the remark this answers is not in the document]
Comment (frankxieke) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsGenerator.java]: Yes, i found other method to calculate if a topic is  persistent topic. Please  have to look.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/LedgerOffloaderMXBeanImpl.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.bookkeeper.mledger.LedgerOffloaderMXBean;
+import static org.apache.bookkeeper.mledger.util.SafeRun.safeRun;
+import org.apache.bookkeeper.mledger.util.StatsBuckets;
+import org.apache.pulsar.common.stats.Rate;
+
+public class LedgerOffloaderMXBeanImpl implements LedgerOffloaderMXBean {
```
[the remark this answers is not in the document]
Comment (frankxieke) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/LedgerOffloaderMXBeanImpl.java:35]: Yes, it is problem.  I will remove the useless topic metrics in the map periodically.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/LedgerOffloaderMXBeanImpl.java [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+    public Map<String, Rate> getReadOffloadErrors() {
+        return readOffloadErrorMap;
+    }
+
+    public void recordOffloadTime(String managedLedgerName, long time, TimeUnit unit) {
+        Rate adder = offloadTimeMap.get(managedLedgerName);
+        if (adder == null) {
+            adder = offloadTimeMap.compute(managedLedgerName, (k, v) -> new Rate());
+        }
+        adder.recordEvent(unit.toMillis(time));
```
[the remark this answers is not in the document]
Comment (frankxieke) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/LedgerOffloaderMXBeanImpl.java]: Yes, there may be race conditions problem. I will fix it. please have a look.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/LedgerOffloaderMXBeanImpl.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        writeToStorageErrorMap.values().forEach(rate->rate.calculateRate(seconds));
+        streamingWriteToStorageRateMap.values().forEach(rate -> rate.calculateRate(seconds));
+        streamingWriteToStorageErrorMap.values().forEach(rate->rate.calculateRate(seconds));
+        readOffloadDataRateMap.values().forEach(rate->rate.calculateRate(seconds));
+        readOffloadErrorMap.values().forEach(rate->rate.calculateRate(seconds));
+        readOffloadIndexLatencyBucketsMap.values().forEach(stat->stat.refresh());
+        readOffloadDataLatencyBucketsMap.values().forEach(stat->stat.refresh());
+    }
+
+    //TODO metrics在namespace这个level的输出。
```
[the remark this answers is not in the document]
Comment (frankxieke) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/LedgerOffloaderMXBeanImpl.java]: It is deleted.

**Review by hangc0276 — CHANGES_REQUESTED** [review]

**Comment by hangc0276** [comment]
I doubt why not choose provide `OpStatsLogger` (pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/DataSketchesOpStatsLogger.java) to generate metrics instead of `LedgerOffloaderMXBeanImpl`.  

An demo of using `OpStatsLogger` is https://github.com/streamnative/kop/pull/389

**Comment by frankxieke** [comment]
rerun failure checks

**Comment by frankxieke** [comment]
rerun failure checks

**Comment by frankxieke** [comment]
> rerun failure checks

rerun failure checks

**Pull request closed without merging**